packages/mcfunction/src/parser/macro.ts

```diff
@@ -0,0 +1,153 @@
+import * as core from '@spyglassmc/core'
+import { localeQuote, localize } from '@spyglassmc/locales'
+import type { MacroChildNode, MacroKeyNode, MacroNode } from '../node/index.js'
+
+/**
+ * Parse a macro.
+ */
+export function macro(): core.Parser<MacroNode> {
+	return (
+		src: core.Source,
+		ctx: core.ParserContext,
+	): core.Result<MacroNode> => {
+		const start = src.cursor
+		const children: MacroChildNode[] = []
+
+		// Handle the starting '$'
+		if (src.peek() == '$' && src.peek(2) !== '$(') {
+			const ans: MacroChildNode = {
+				type: 'mcfunction:macro_child',
+				range: core.Range.create(start, start + 1),
+				options: { type: 'sign' },
+				value: '$',
+			}
+			src.skip()
+			children.push(ans)
+		}
+
+		// Handle the rest of the line
+
+		let beginning = src.cursor
+		let txt = src.readUntil('$', '\r', '\n')
+		do { // Do-while because we always want it to work the first time
+			let wasMacro = false
+			const child: MacroChildNode = parse(src, ctx, txt, beginning, wasMacro)
+			if (child.options.type === 'macro') {
+				wasMacro = true
+			}
+			// Handle a non-macro $ in the middle of the line
+			let subMacro: MacroChildNode | undefined
+			if (
+				child.options.type === 'other' && child.value?.indexOf('$') !== -1
+			) {
+				const parts: String[] = child.value?.split('$') ?? []
```

Comment:
Why is this splitting the string result of the previous `parse` call? What would the parts even be in a normal command?

Reply:
In a normal command it wouldn't be split because it wouldn't contain a `$`, it's designed only to handle that edge case